src/main/java/net/minecraftforge/resource/ResourceCacheManager.java

```diff
@@ -0,0 +1,375 @@
+        private static final Logger LOGGER = LogUtils.getLogger();
+
+        /**
+         * The pack type this manager is responsible for.
+         */
+        private final PackType packType;
+        /**
+         * The namespace this manager is responsible for.
+         */
+        private final String namespace;
+        /**
+         * Indicates if the indexing should run off-thread.
+         */
+        private final boolean indexOffThread;
+        /**
+         * The path builder to use.
+         */
+        private final BiFunction<PackType, String, Path> pathBuilder;
+        /**
+         * The path walker stream factory to use.
+         */
+        private final PathWalkerFactory pathFinder;
+
+        /**
+         * The cache entries for this manager.
+         */
+        private final Map<String, List<ResourceCacheEntry>> entriesByPathPrefix = Maps.newConcurrentMap();
+        /**
+         * Indicates if the cache has been loaded successfully.
+         */
+        private final AtomicBoolean cacheLoaded = new AtomicBoolean(false);
+
+        /**
+         * Creates a new namespaced resource cache manager.
+         *
+         * @param packType       The pack type that this manager handles.
+         * @param namespace      The namespace that this manager handles.
+         * @param indexOffThread True to enable indexing off-thread.
+         * @param pathBuilder    The path builder to use.
+         * @param pathFinder     The path walker stream factory to use.
+         */
+        private NamespacedResourceCacheManager(final PackType packType, final String namespace, final boolean indexOffThread, BiFunction<PackType, String, Path> pathBuilder, final PathWalkerFactory pathFinder) {
+            this.packType = packType;
+            this.namespace = namespace;
+            this.indexOffThread = indexOffThread;
+            this.pathBuilder = pathBuilder;
+            this.pathFinder = pathFinder;
+        }
+
+        /**
+         * Triggers indexing of this manager.
+         */
+        public void index() {
+            if (indexOffThread) {
+                //Run off-thread.
+                CompletableFuture.runAsync(
+                        this::doIndex,
+                        Util.backgroundExecutor()
+                );
+            } else {
+                //Run on-thread
+                doIndex();
+            }
+        }
+
+        /**
+         * Performs the actual indexing of this manager.
+         * Can be invoked on- or off-thread.
+         */
+        private void doIndex() {
+            //Get the path to the root of the namespace in the current pack.
+            final Path rootPath = pathBuilder.apply(packType, namespace);
+
+            //Build a walkable stream, process it
+            try (final Stream<Path> paths = pathFinder.createWalkingStream(rootPath)) {
+                paths
+                    .parallel() //Run the stream in parallel
```

Comment:
I feel like the initial call which just makes it parallel can be on the same line as `paths`